File: dsa/cpp/week_3/03_linked_list_cycle.cpp
// Key Takeaways:
// 1. Two approaches: hash set (O(n) space) stores visited node pointers,
//    Floyd's cycle detection (O(1) space) uses fast/slow pointers.
//    Always prefer Floyd's when space is constrained.
//
// 2. In C++, use unordered_set<ListNode*> for O(1) average lookup.
//    unordered_set uses a hash table; set uses a red-black tree (O(log n)).
//    For pointer storage with no ordering needed, unordered_set is correct.
//
// 3. Floyd's: slow moves 1 step, fast moves 2. If they meet, cycle exists.
//    Always check fast != nullptr && fast->next != nullptr before advancing
//    to avoid dereferencing a null pointer (segfault).
//
// 4. This same Floyd's algorithm works on arrays (LeetCode #287) by treating
//    array values as next pointers — same pattern, different data structure.
//
// Complexity: Hash set — O(n) time, O(n) space. Floyd's — O(n) time, O(1) space.

// Compile: g++ -std=c++17 -o 03_linked_list_cycle 03_linked_list_cycle.cpp
#include <iostream>
#include <vector>
#include <string>
#include <set>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

bool hasCycleSet(ListNode* head) {
    // Your implementation
    set<ListNode*> seen;
    
    for (ListNode* ptr = head; ptr != nullptr; ptr = ptr->next) {
        if (seen.find(ptr) == seen.end()) {
            seen.insert(ptr);
        }
        else {
            return true;
        }
    }
    
    return false;
}

bool hasCycleFloyd(ListNode* head) {
    // Your implementation
    ListNode* slow = head;
    ListNode* fast = head;

    while (fast != nullptr && fast->next != nullptr) {
        slow = slow->next;
        fast = fast->next->next;

        if (slow == fast) {
            return true;
        }
    }

    return false;
}

ListNode* makeListWithCycle(const vector<int>& values, int cyclePos) {
    if (values.empty()) return nullptr;
    vector<ListNode*> nodes;
    for (int v : values) nodes.push_back(new ListNode(v));
    for (size_t i = 0; i < nodes.size() - 1; i++) nodes[i]->next = nodes[i + 1];
    if (cyclePos >= 0 && cyclePos < (int)nodes.size()) nodes.back()->next = nodes[cyclePos];
    return nodes[0];
}

vector<int> makeRange(int n) {
    vector<int> v(n);
    for (int i = 0; i < n; i++) v[i] = i;
    return v;
}

struct TestCase { vector<int> values; int cyclePos; bool expected; string desc; };

int main() {
    vector<TestCase> tests = {
        {{3, 2, 0, -4}, 1, true, "cycle at pos 1"},
        {{1, 2}, 0, true, "cycle at head"},
        {{1, 2, 3, 4}, 0, true, "cycle back to head"},
        {{1, 2, 3, 4}, 2, true, "cycle in middle"},
        {{1}, -1, false, "single node no cycle"},
        {{1, 2}, -1, false, "two nodes no cycle"},
        {{1, 2, 3, 4, 5}, -1, false, "five nodes no cycle"},
        {{}, -1, false, "empty list"},
        {{1}, 0, true, "self-loop"},
        {{1, 2, 3}, -1, false, "three nodes no cycle"},
        {makeRange(100), -1, false, "large list no cycle"},
        {makeRange(100), 50, true, "large list cycle at 50"},
        {makeRange(100), 0, true, "large list cycle at head"},
        {makeRange(100), 99, true, "large list self-loop tail"},
        {{1, 2, 3, 4, 5}, 4, true, "tail self-loop"},
        {{1, 2, 3, 4, 5}, 3, true, "cycle at second to last"},
        {{1, 2, 3, 4, 5}, 1, true, "cycle at pos 1 five nodes"},
        {{-1, -2, -3}, -1, false, "negative values no cycle"},
        {{-1, -2, -3}, 0, true, "negative values with cycle"},
        {{1, 2}, 1, true, "two nodes tail self-loop"},
    };

    struct Approach { string name; bool (*fn)(ListNode*); };
    vector<Approach> approaches = {{"HashSet", hasCycleSet}, {"Floyd's", hasCycleFloyd}};

    cout << "======================================================================" << endl;
    cout << "LINKED LIST CYCLE - Test Results" << endl;
    cout << "======================================================================" << endl;

    int total = tests.size();
    for (auto& a : approaches) {
        cout << "\n" << a.name << ":" << endl;
        int passed = 0;
        for (size_t i = 0; i < tests.size(); i++) {
            auto& tc = tests[i];
            ListNode* head = makeListWithCycle(tc.values, tc.cyclePos);
            bool result = a.fn(head);
            bool ok = (result == tc.expected);
            if (ok) passed++;
            cout << "  " << (i + 1) << ". [" << (ok ? "PASS" : "FAIL") << "] " << tc.desc
                 << ": pos=" << tc.cyclePos << " -> " << (result ? "true" : "false") << endl;
        }
        cout << "  Result: " << passed << "/" << total << endl;
    }

    cout << "\n======================================================================" << endl;
    cout << "Questions:" << endl;
    cout << "1. Why does Floyd's guarantee fast meets slow?" << endl;
    cout << "2. Hash set vs Floyd's trade-off?" << endl;
    cout << "3. Find WHERE cycle begins (LeetCode #142)?" << endl;
    return 0;
}

File: dsa/python/week_3/03_linked_list_cycle.py
# Key Takeaways:
# 1. Two approaches: hash set (O(n) space) stores visited nodes and checks membership,
#    Floyd's cycle detection (O(1) space) uses fast/slow pointers — if they ever meet,
#    there's a cycle. Always prefer Floyd's when space is constrained.
#
# 2. Floyd's invariant: slow moves 1 step, fast moves 2 steps. If a cycle exists,
#    fast will lap slow and they will meet inside the cycle. If no cycle, fast
#    reaches None first.
#
# 3. In Python, use a set() to store visited node references (object identity).
#    `node in seen` checks if the exact object was seen before, not the value.
#
# 4. This same Floyd's algorithm detects cycles in arrays too (LeetCode #287
#    Find Duplicate Number) — treating array values as "next pointers".
#
# Complexity: Hash set — O(n) time, O(n) space. Floyd's — O(n) time, O(1) space.

"""
DSA Problem: Linked List Cycle

Tests: Floyd's cycle detection, two pointers (fast/slow), space optimization

Difficulty: Easy
Source: LeetCode #141

Problem:
Given head of a linked list, determine if it has a cycle.

Constraints:
    - Number of nodes in range [0, 10^4]
    - -10^5 <= Node.val <= 10^5
"""

from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: "ListNode" = None):
        self.val = val
        self.next = next


def has_cycle_set(head: Optional[ListNode]) -> bool:
    """
    Hash set approach - track visited nodes.

    Time: O(?)
    Space: O(?)

    Your implementation:
    """
    visited = set()

    while head != None:
        if head not in visited:
            visited.add(head)
            head = head.next
        else:
            return True
        
    
    return False



def has_cycle_floyd(head: Optional[ListNode]) -> bool:
    
    """
    Floyd's tortoise and hare - O(1) space. fast and slow pointer

    Time: O(?)
    Space: O(?)

    Your implementation:
    """
    if not head or not head.next:
        return False
    slow = head
    fast = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
        

    return False


# =============================================================================
# Helper functions
# =============================================================================
def make_list_with_cycle(values: list, cycle_pos: int) -> Optional[ListNode]:
    """Create a linked list. cycle_pos = index where tail connects back (-1 = no cycle)."""
    if not values:
        return None
    nodes = [ListNode(v) for v in values]
    for i in range(len(nodes) - 1):
        nodes[i].next = nodes[i + 1]
    if cycle_pos >= 0:
        nodes[-1].next = nodes[cycle_pos]
    return nodes[0]


# =============================================================================
# Test Cases - LeetCode Level
# =============================================================================
if __name__ == "__main__":
    test_cases = [
        {"values": [3, 2, 0, -4], "cycle_pos": 1, "expected": True, "desc": "cycle at pos 1"},
        {"values": [1, 2], "cycle_pos": 0, "expected": True, "desc": "cycle at head"},
        {"values": [1, 2, 3, 4], "cycle_pos": 0, "expected": True, "desc": "cycle back to head"},
        {"values": [1, 2, 3, 4], "cycle_pos": 2, "expected": True, "desc": "cycle in middle"},
        {"values": [1], "cycle_pos": -1, "expected": False, "desc": "single node no cycle"},
        {"values": [1, 2], "cycle_pos": -1, "expected": False, "desc": "two nodes no cycle"},
        {"values": [1, 2, 3, 4, 5], "cycle_pos": -1, "expected": False, "desc": "five nodes no cycle"},
        {"values": [], "cycle_pos": -1, "expected": False, "desc": "empty list"},
        {"values": [1], "cycle_pos": 0, "expected": True, "desc": "self-loop"},
        {"values": [1, 2, 3], "cycle_pos": -1, "expected": False, "desc": "three nodes no cycle"},
        {"values": list(range(100)), "cycle_pos": -1, "expected": False, "desc": "large list no cycle"},
        {"values": list(range(100)), "cycle_pos": 50, "expected": True, "desc": "large list cycle at 50"},
        {"values": list(range(100)), "cycle_pos": 0, "expected": True, "desc": "large list cycle at head"},
        {"values": list(range(100)), "cycle_pos": 99, "expected": True, "desc": "large list self-loop tail"},
        {"values": [1, 2, 3, 4, 5], "cycle_pos": 4, "expected": True, "desc": "tail self-loop"},
        {"values": [1, 2, 3, 4, 5], "cycle_pos": 3, "expected": True, "desc": "cycle at second to last"},
        {"values": [1, 2, 3, 4, 5], "cycle_pos": 1, "expected": True, "desc": "cycle at pos 1"},
        {"values": [-1, -2, -3], "cycle_pos": -1, "expected": False, "desc": "negative values no cycle"},
        {"values": [-1, -2, -3], "cycle_pos": 0, "expected": True, "desc": "negative values with cycle"},
        {"values": [1, 2], "cycle_pos": 1, "expected": True, "desc": "two nodes tail self-loop"},
    ]

    print("=" * 70)
    print("LINKED LIST CYCLE - Test Results")
    print("=" * 70)

    for approach_name, func in [("HashSet", has_cycle_set), ("Floyd's", has_cycle_floyd)]:
        print(f"\n{approach_name}:")
        passed = 0
        total = len(test_cases)
        for i, tc in enumerate(test_cases, 1):
            head = make_list_with_cycle(tc["values"], tc["cycle_pos"])
            result = func(head)
            ok = result == tc["expected"]
            passed += ok
            status = "PASS" if ok else "FAIL"
            vals = tc["values"] if len(tc["values"]) <= 8 else tc["values"][:5] + ["..."]
            print(f"  {i:2}. [{status}] {tc['desc']}: vals={vals}, pos={tc['cycle_pos']} -> {result}")
        print(f"  Result: {passed}/{total}")

    print("\n" + "=" * 70)
    print("Questions:")
    print("1. Why does Floyd's algorithm guarantee the fast pointer meets the slow pointer?")
    print("2. What's the trade-off between the hash set and Floyd's approach?")
    print("3. Follow-up: How would you find WHERE the cycle begins? (LeetCode #142)")
